### ISSUE TITLE
Add CONNECT support

### ISSUE THREAD

**Issue by mattwoodyard** [envoyproxy/envoy#1451]
A follow up to #1214. Add support for CONNECT to envoy. There are 2 use cases for CONNECT that I'd like to add. Envoy receiving a CONNECT request, connecting to an upstream, issuing a 200 and then downgrading to L4 (similar to websocket support). The second is envoy using CONNECT when initiating an upstream L4 connection (the moral equivalent of envoy using http_proxy for some https upstreams). I think the first one is pretty straightforward. 

- Add a config option to enable CONNECT on the listener in http1_options
- Update codec to support the right authority in CONNECT
- Write a Http11ConnectHandlerImpl in the spirit of WsHandlerImpl

The second case is a little more unclear to me. Here's a straw man. 

- Upstreams get a 'use connect' flag.
- When enabled any connections to that upstream are initiated with a CONNECT request exchange before the remain traffic is forwarded.

**Comment by mattklein123** [comment]
Both sound reasonable to me. For the upstream case, I assume we just need to support HTTP/1? I think this could be folded into the http1_codec_options which should be available to upstream clusters in v1 and v2 APIs.

**Comment by mattklein123** [comment]
> Write a Http11ConnectHandlerImpl in the spirit of WsHandlerImpl

Please share as much of the code as possible, but yes, I think the flow will be very similar.

**Comment by jippi** [comment]
I would love to see support in evnoy for https://linkerd.io/features/http-proxy/

**Comment by mattklein123** [comment]
@jippi http_proxy is orthogonal to CONNECT. Envoy does support basic http_proxy support, but it's on by default. See the "allow_absolute_url" option here: https://www.envoyproxy.io/docs/envoy/latest/api-v1/network_filters/http_conn_man (and similar v2 option). 

What is not supported is DNS lookup for previously unknown backends. This is tracked in https://github.com/envoyproxy/envoy/issues/1606.

This could use better coverage in arch overview (https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/http_connection_management) if you feel like doing a doc PR.

**Comment by pschultz** [comment]
Please note that the default Go HTTP client [looks at HTTP_PROXY for HTTPS requests](https://golang.org/pkg/net/http/#ProxyFromEnvironment). HTTPS_PROXY takes precedence, but if that's empty HTTP_PROXY is in effect. This means we cannot blindly set HTTP_PROXY to an Envoy listener for all clients right now.

We currently work around this by replacing the Proxy function in the http.DefaultTransport (so https requests don't pass through Envoy), which isn't great.

**Comment by salrashid123** [comment]
I'm interested in envoy support for CONNECT specifically to see how it can help [istio's egress rules](https://istio.io/docs/tasks/traffic-management/egress.html)  which at the moment requires rewriting _at the client_ ```https://``` --> ```http://...:443``` (for example).

this FR should help with that in that if the running container in istio-k8s has an env setting, it'd 'speak' to envoy via CONNECT for https requests (or transparently)

 I played around with a local config (without istio or envoy for that matter), the http_proxy= locally would be something like this:

I made up a sample app 
```yaml
    spec:
      containers:
      - name: myapp-container
        image: salrashid123/my_app
        ports:
        - containerPort: 8080
        env:
        - name: HTTP_PROXY
          value: "http://localhost:3128"
      - name: sslproxy
        image: salrashid123/sslproxy
        ports:
        - containerPort: 3128
```
(w/ istio in the picture the proxy is envoy itself)

I used squid which allows for really advanced content filtering (i.e if you happens to trust a CA squid provides, you can even inspect the transit traffic and make determinations based on some signal there; here's [my example](https://github.com/salrashid123/squid_proxy/tree/master/content_adaptation) )....but i'd be happy w/ envoy and CONNECT if it allows istio to set simple host based rule (i.,e allow https traffic to foo.com but not.bar.com)

**Comment by sean-brandt** [comment]
This is also very important for a lot of corporate usecases where the destination may not be 'local' to where envoy is running. Having CONNECT support on a per destination basis is really key in those situations where corporate requirements have forward proxies ( non-transparent ) in place to transit security boundaries.

**Comment by jcampbell05** [comment]
We have written a very small test proxy to implement this situation (http://github.com/aiwip/super-proxy) however we think it would be much better to use a well tested and used L7 proxy.

What is the best way to implement this ?

**Comment by mattklein123** [comment]
The plan for implementing this is going to be:
1) Finish the ongoing dynamic forward proxy support
2) Implement CONNECT in the forward proxy filter using upgrade semantics, similar to websocket.

**Comment by jespersoderlund** [comment]
Will this also support responding to a CONNECT from an application?

**Comment by mattklein123** [comment]
> Will this also support responding to a CONNECT from an application?

What do you mean exactly?

**Comment by mattklein123** [comment]
@alyssawilk I started to look briefly at implementing this. I think roughly it will involve:
1) Making sure we allow CONNECT to propagate into the filter chain (probably via an HCM config setting to allow CONNECT)
2) We likely need to modify the route matching config to handle no path and CONNECT
3) Then we can have the dynamic_proxy_filter handle CONNECT and use the tcp_proxy code to initiate outbound connections.

I'm no longer thinking that we need to involve any of the upgrade config in this. WDYT? Any other thoughts about the right way to approach this from a config perspective?

**Comment by alyssawilk** [comment]
1) we definitely need configuration for connect.  I do think it's worth reusing the existing upgrade configs for it, because I can imagine folks wanting a different filter chain for it, given the payload is not really HTTP body
2) +10 on no path.  Also general audit of other code?
3) why tcp proxy mode?  AFIK (you can check with piotr) CONNECT is legal for H2, so I'd think we'd want to use standard upstream as we do for ugprades.

**Comment by mattklein123** [comment]
> I do think it's worth reusing the existing upgrade configs for it, because I can imagine folks wanting a different filter chain for it, given the payload is not really HTTP body

How do you suggest configuring it? Should we have a special upgrade type of "CONNECT"? That seems pretty hacky. Do you prefer a specific boolean for CONNECT?

> why tcp proxy mode? AFIK (you can check with piotr) CONNECT is legal for H2, so I'd think we'd want to use standard upstream as we do for ugprades.

My understanding of CONNECT is that we start a raw TCP connection to the upstream in this case (that is my read of the spec). cc @PiotrSikora

**Comment by mattklein123** [comment]
Maybe just introduce an independent `ConnectConfig` which would also turn on CONNECT handling? That seems the safest?

**Comment by mattklein123** [comment]
(Then since we have will have route matching on CONNECT it seems like a straightforward way to either use the connect config or not.)

**Comment by alyssawilk** [comment]
I'm fine with separate config, if that's how we want to handle it.  I think we're going to want to handle both proxying the connect (to be handled on a later hop) and handling the connect.   I guess for forward proxy you're only concerned about the latter, where using TCP does make sense.  I think my eventual use case is more about the former (which is why I was thinking HTTP backend) .  Let's just consider designing APIs and code to easily do both in future, yeah?

**Comment by PiotrSikora** [comment]
> 3. why tcp proxy mode?  AFIK (you can check with piotr) CONNECT is legal for H2, so I'd think we'd want to use standard upstream as we do for ugprades.

> My understanding of CONNECT is that we start a raw TCP connection to the upstream in this case (that is my read of the spec). cc @PiotrSikora

You're both correct.

In HTTP/1.1 and HTTP/2 the `CONNECT` method is used to ask proxy to open a raw TCP connection to the requested authority.

However, the extended `CONNECT` method introduced in [RFC8441](https://tools.ietf.org/html/rfc8441) allows tunneling TCP connections over HTTP/2 using `HEADERS` and `DATA` frames.

**Comment by mattklein123** [comment]
OK sounds good I will make sure the connect config is suitable for both "normal" CONNECT and "extended" CONNECT even though initially I will likely only implement "normal" CONNECT.

**Comment by jcampbell05** [comment]
In my use case I want to be able to have Envoy consume a `CONNECT` connection from a client and act as a MITTM reverse-forward proxy so I can filter the connection strip the TLS and redirect it to another server of my choice.

I know the security implications of this but I have a very good reason for doing this.

**Comment by jcampbell05** [comment]
> What do you mean exactly?

I think they mean will it support handling connections from a client application i.e an iOS app doing a CONNECT request to a envoy proxy instance to be handled

**Comment by emaildanwilson** [comment]
It looks like this feature is what we would need to configure a tier of envoy servers which would be used as a proxy pool for systems which are unable to have an envoy sidecar and also not change the way they make requests (ie without setting host headers or changing the servers they reach with host file/dns manipulation, etc.). Is that correct?

We can pony up a resource to contribute this feature unless someone is already working on it.

@mattklein123

**Comment by mattklein123** [comment]
@emaildanwilson I'm not working on this and won't be working on it soon. If you have someone to work on it please have them reach out and I can guide them.

**Comment by cetezadi** [comment]
I am going to start in on this per our earlier conversation.  It'll be slow, but hopefully beneficial.

**Comment by jespersoderlund** [comment]
> I am going to start in on this per our earlier conversation. It'll be slow, but hopefully beneficial.

Any progress on this? Can we help out in testing this?

**Comment by mattklein123** [comment]
@alyssawilk might end up working on this during the quarter as she is working on something related, so please ping this issue before starting work if anyone else plans on working on this.

**Comment by mattklein123** [comment]
>  but I think the cleanest place would be in the new TCP upstream, which could do standard push-back buffering until it got the response (where the server codec would have do something klunkier with special case buffering the direct dispatch path which is already pretty complicated. Thoughts?

Yes big +1 for doing it in the router/upstream handler so we can do standard flow control. I think it would be OK also to add this as an improvement in a follow up for the alpha support, but up to you if it's harder or easier to do it after the fact.

**Comment by chradcliffe** [comment]
Are there any plans to add support in this first pass for initiating an upstream CONNECT request independent of whether the downstream request was a CONNECT request? In the original issue this was called the "second case", but I haven't seen much mention of it lately.

We have a use case for initiating upstream CONNECT requests, so it would be helpful to know if this is work that hasn't been picked up by anyone else yet.

**Comment by mattklein123** [comment]
> Are there any plans to add support in this first pass for initiating an upstream CONNECT request independent of whether the downstream request was a CONNECT request? In the original issue this was called the "second case", but I haven't seen much mention of it lately.

Is the use case here to have tcp_proxy have an upstream CONNECT mode? I'm guessing @alyssawilk is not working on that right now but it seems not enormously difficult after all of the ongoing work is done.

**Comment by chradcliffe** [comment]
That would be one use case.

Another use case would be to proxy downstream plain HTTP as upstream HTTPS via another upstream proxy. A request path would look something like this:

`Client --[plain HTTP]--> Envoy --[HTTP CONNECT]--> Upstream Proxy --[HTTPS]--> Server`

I hope that's clear enough.

It seems like that might be quite a bit more complicated than the tcp_proxy case, though.

**Comment by mattklein123** [comment]
> It seems like that might be quite a bit more complicated than the tcp_proxy case, though.

It actually might not be that complicated if @alyssawilk follows though and has code that will wait for the upstream CONNECT response (we could initiate one instead). It will definitely be extra work though. I would suggest looping back once the initial support lands, though hopefully @alyssawilk can keep this in mind while she is implementing.

**Comment by alyssawilk** [comment]
I already added support for sending TCP over HTTP/2 CONNECT (https://github.com/envoyproxy/envoy/issues/1630) - it's missing the all important proxy proto header but I plan on landing that in a follow-up thanks to @wez470 's handy utilities and that should handle the Envoy-to-Envoy case as described in https://github.com/envoyproxy/envoy/issues/1630

Would that work for you or do you have some non-Envoy servers in the line?  I hadn't planned on implementing TCP->CONNECT for HTTP/1 (since we prefer HTTP/2 multiplexing in-house) but I'd be happy to walk someone through the steps of adding if you need it.

**Comment by chradcliffe** [comment]
We use some upstream non-envoy servers that only support HTTP/1 CONNECT, so HTTP/2 CONNECT wouldn't work for our use case.

@alyssawilk thank you for the offer to walk someone through the steps -- we're planning to start on this functionality over the next month or two, so as soon we know who is going to be doing the work I'll get in touch.

**Comment by alyssawilk** [comment]
Sounds good.  FWIW it should be fairly minimal coding - I'd expect it to be  90% writing test cases making sure that HTTP/1.1 codec logic around chunks doesn't get screwy, ideally most of which you can copy-paste-edit from other PRs on these two issues :-)

**Comment by mabukhovsky** [comment]
Hi @alyssawilk, I've see you've pushed several PRs to add CONNECT support. Could you help me to understand - with the current state of master branch, will I be able to resolve [my issue](https://github.com/envoyproxy/envoy/issues/10160#issuecomment-591432025)? 
```
$ curl --header "X-Host-Port: www.yahoo.com" https://www.google.com
curl: (56) Received HTTP code 403 from proxy after CONNECT
$ curl --header "X-Host-Port: www.yahoo.com:443" https://www.google.com
curl: (56) Received HTTP code 403 from proxy after CONNECT
$ curl --header "X-Host-Port: www.yahoo.com:80" https://www.google.com
curl: (56) Received HTTP code 403 from proxy after CONNECT
```

**Comment by alyssawilk** [comment]
Hey @mabukhovsky 
So I'm unfamiliar with the internals of what curl is doing here.  Just going based on the logs, if it's using a CONNECT, I'd imagine that if you properly configured Envoy after #10623 lands it might help your issue.  I'll land some sample configs in follow-up PRs but you can crib off the intergration tests right now for how to enable connect support (as it's off by default) as well as how to turn on connect termination (it proxies by default)

**Comment by alyssawilk** [comment]
At this point, CONNECT support (both proxying the CONNECT request, or terminating CONNECT (stripping the CONNECT headers and sending raw TCP upstream) are implemented and are ready for testing.  I would suggest not using it production until #10974 lands.

Docs exist (but are hidden) and sample config coming soon (currently you can crib from the integration tests).

In the next few weeks we should get in no particular order
0) the security improvements in #10974
1) example configs, for both forms of connect as well as tunneling over HTTP/2
2) proxy proto support
4) fixing the current hack where a synthetic path header is added to pathless CONNECT requests
Once all these land I'll unhide the docs, and say we're in stable alpha. The sooner you play around the sooner you can tell me if the config set up works for you though, so feel free to start playing around.

**Comment by chradcliffe** [comment]
I've been experimenting with the terminating CONNECT support on master and noticed some unexpected behaviour. The 200 response from Envoy does not get sent until actual TCP data is sent from the upstream. I thought it was an issue with my test setup, but then I decided to actually read the docs and found the following in the [section on HTTP upgrades](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/http/upgrades.html):

> On receipt of initial TCP data from upstream, the router will synthesize 200 response headers, and then forward the TCP data as the HTTP response body.

So it appears that this is working as intended. My issue is that I can't get a flow like this to work with the current support:

```
Client -> Envoy : HTTP CONNECT
Envoy <-> Upstream : TCP handshake
Envoy -> Client : HTTP 200 OK // this is never sent in the current implementation
Client -- (via Envoy) --> Upstream : HTTP GET
```

The following, however, does work in the current implementation:

```
Client -> Envoy : HTTP CONNECT with tunnelled HTTP GET request
Envoy <-> Upstream : TCP handshake
Envoy -> Upstream : tunnelled HTTP GET request
Upstream - Envoy -> : tunnelled HTTP response
Envoy -> Client : HTTP 200 OK
Envoy -> Client : tunnelled HTTP response
```

But the [HTTP RFC](https://tools.ietf.org/html/rfc7231#section-4.3.6) warns against sending payload with a HTTP CONNECT request:

>  A payload within a CONNECT request message has no defined semantics; sending a payload body on a CONNECT request might cause some existing implementations to reject the request.

Python's `http.client.HTTPConnection` CONNECT support, for example, [sends headers only](https://github.com/python/cpython/blob/master/Lib/http/client.py#L884) on the request by default, so afaik the current CONNECT support won't be compatible with that client.

Is there a plan to implement synthesis of the 200 response after the upstream handshake but prior to seeing upstream bytes? Is there an implementation detail that makes this more difficult than the current behaviour?

**Comment by alyssawilk** [comment]
Oh interesting, so in this case you need the payload data to be sent from the client, in order to trigger the server to send a response which causes the 200 headers?
Yeah, I can fix that.  Woo alpha testers!  :-)

**Comment by chradcliffe** [comment]
So I think ideally the 200 response should be triggered by the upstream TCP handshake completion. The clients will typically wait on the response before sending any tunnelled payload.

**Comment by mattklein123** [comment]
Going to call this done. Let's track bugs / features in new issues. Thank you @alyssawilk!!!

**Comment by davidkarlsen** [comment]
I'm very interested in the 2nd usecase with Istio. Is this possible - and how do I configure that?

**Comment by alyssawilk** [comment]
cc @lambdai
